persistence/src/main/java/com/epam/esm/dao/impl/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,215 @@
+            "join CertificateDetails on GiftCertificates.ID = CertificateDetails.CertificateID " +
+            "join Tags on Tags.ID = CertificateDetails.TagID " +
+            "where CertificateDetails.CertificateID in " +
+            "( " +
+            "select CertificateDetails.CertificateID from CertificateDetails " +
+            "join Tags on Tags.ID = CertificateDetails.TagID " +
+            "where Tags.name = ?" +
+            ")";
+
+    private static final String SQL_ADD_CERTIFICATE =
+            "insert into GiftCertificates " +
+            "(Name, Description, Price, CreateDate, LastUpdateDate, Duration) " +
+            "values (?, ?, ?, ?, ?, ?)";
+
+    private static final String SQL_DELETE_CERTIFICATE =
+            "delete from GiftCertificates where (ID = ?)";
+
+    private static final String SQL_UPDATE_CERTIFICATE =
+            "update GiftCertificates " +
+            "set Name = IFNULL(?, Name), Description = IFNULL(?, Description)," +
+            "Price = IFNULL(?, Price), LastUpdateDate = IFNULL(?, LastUpdateDate), " +
+            "Duration = IFNULL(?, Duration) " +
+            "where (ID = ?)";
+
+    private static final String SQL_CREATE_JOIN =
+            "insert into CertificateDetails (CertificateID, TagID) values (?, ?)";
+
+    private static final String SQL_DELETE_JOIN =
+            "delete from CertificateDetails where (CertificateID = ?)";
+
+    private final JdbcTemplate jdbcTemplate;
+    private final ResultSetExtractor<List<GiftCertificate>> extractor;
+    private final RowMapper<GiftCertificate> mapper;
+
+    @Autowired
+    public GiftCertificateDaoImpl(
+            JdbcTemplate jdbcTemplate, ResultSetExtractor<List<GiftCertificate>> extractor,
+            RowMapper<GiftCertificate> mapper) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.extractor = extractor;
+        this.mapper = mapper;
+    }
+
+    @Override
+    public GiftCertificate getGiftCertificate(String name) {
+        return jdbcTemplate.queryForObject(SQL_GET_CERTIFICATE_BY_NAME, mapper, name);
+    }
+
+    @Override
+    public GiftCertificate getGiftCertificate(int id) {
+        return jdbcTemplate.queryForObject(SQL_GET_CERTIFICATE_BY_ID, mapper, id);
+    }
+
+    @Override
+    public List<GiftCertificate> getAllGiftCertificates() {
+        return jdbcTemplate.query(SQL_GET_ALL_CERTIFICATES, extractor);
+    }
+
+    @Override
+    public List<GiftCertificate> getAllGiftCertificates(String content) {
+        String param = "%" + content + "%";
+        return jdbcTemplate.query(
+                SQL_GET_ALL_CERTIFICATES_BY_CONTENT, extractor, param, param);
+    }
+
+    @Override
+    public List<GiftCertificate> getGiftCertificateByTagName(String tagName) {
+        return jdbcTemplate.query(SQL_GET_CERTIFICATES_BY_TAG_NAME, extractor, tagName);
+    }
+
+    @Override
+    public List<GiftCertificate> getAllGiftCertificatesSortedByName(boolean isAscending) {
```

Comment:
> Do sorting in service layer

I have done sorting in a database query, because the database is optimized for search and sorting, and in the tests,  database showed a better sorting result than getting all records from the database and then sorting them on server. This is the practice that Svyatoslav Svyatoslavovich Kulikov advised, including in his book *work with MYSQL, MS SQL SERVER and ORACLE in examples*.